### PR TITLE
[GSIP-169] WPS Extension "GetExecutions" Operation

### PR DESCRIPTION
**Pull request by afabiani**
ref to https://github.com/geoserver/geoserver/wiki/GSIP-169

### CODE REVIEW

**Comment by tbarsballe** [comment]
This has missed the window to be included in 2.14.0, so the proposal should probably updated to say **Assigned to Release** - This proposal is for GeoServer 2.15. (Currently it says 2.14)

**Comment by afabiani** [comment]
"activemq" module is failing Travis build for unknown reasons

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/GetExecutionsTransformer.java [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+                                                }
+                                                end("wfs:GetFeature");
+                                            } else {
+                                                element(
+                                                        "wps:ComplexData",
+                                                        "<![CDATA[" + out.toString() + "]]",
+                                                        inputAttributes);
+                                            }
+                                        } catch (Exception e) {
+                                            // LOGGER.log(Level.WARNING, "", e);
```
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/GetExecutionsTransformer.java]: Please don't eat exceptions. The formatting should also suggest you to roll the encoding of the specific inputs into a separate method, and do a catch around it.

File: src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusPage.java [hunk 2]
```diff
@@ -46,6 +51,15 @@ protected Component getComponentForProperty(
                             IModel<ExecutionStatus> itemModel,
                             Property<ExecutionStatus> property) {
                         // have the base class create a label for us
+                        Object value = property.getPropertyValue(itemModel.getObject());
+                        if (value instanceof Date) {
+                            SimpleDateFormat gmtFrmt =
```
[the remark this answers is not in the document]
Comment (afabiani) [reply src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusPage.java]: right, I have treated it as a plain label and removed datetime deps from pom.xml

File: src/extension/wps/web-wps/src/test/java/org/geoserver/wps/web/ProcessStatusPageTest.java [hunk 3]
```diff
@@ -55,11 +63,28 @@ public void test() throws Exception {
...
         // start the page, should have one process running
         tester.startPage(new ProcessStatusPage());
         // print(tester.getLastRenderedPage(), true, true);
         tester.assertLabel("table:listContainer:items:1:itemProperties:3:component", "gs:Monkey");
         tester.assertLabel("table:listContainer:items:1:itemProperties:5:component", "RUNNING");
+        List<ExecutionStatus> executions = getItems();
+        assertEquals(1, executions.size());
+        ExecutionStatus status = executions.get(0);
+        DateFormat df =
+                new SimpleDateFormat(
```
[the remark this answers is not in the document]
Comment (afabiani) [reply src/extension/wps/web-wps/src/test/java/org/geoserver/wps/web/ProcessStatusPageTest.java]: done, thanks.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/Executions.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        this.ctx = ctx;
+    }
+
+    public Object run(GetExecutionsType request) {
+        // Check whether the user is authenticated or not and, in the second case, if it is an
+        // Administrator or not
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        final String principal =
+                auth != null && auth.getPrincipal() != null ? auth.getPrincipal().toString() : null;
+        boolean isAdmin = getSecurityManager().checkAuthenticationForAdminRole(auth);
```
[the remark this answers is not in the document]
Comment (afabiani) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/Executions.java]: Currently, it is not dynamically generated. I'm just getting the security context from the thread local.

**Comment by afabiani** [comment]
@aaime I have taken action accordingly to your review. Thanks a lot for your time.

File: src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java [hunk 5]
```diff
@@ -179,16 +181,17 @@ private Filter getFullSearch(String[] keywords) {
...
             query.setMaxFeatures((int) count);
         }
-        SortParam sort = getSort();
+
+        final SortParam sort = getSort();
+
+        SortBy sortOrder = null;
         if (sort != null) {
-            SortByImpl[] sortBys = new SortByImpl[1];
+            SortBy[] sortBys = new SortByImpl[1];
```
Comment (aaime) [inline src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java]: Still having a StorByImpl around ;-)

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessListenerNotifier.java [hunk 6]
```diff
@@ -64,6 +66,24 @@ public void fireProgress(float progress, String task) {
             }
             status.setProgress(progress);
             status.setTask(task);
+
+            // Update the estimated completion.
+            // By default we estimate the completion as:
+            // "time elapsed / percentage completed"
+            if (progress > 0) {
+                long timeEalpsedMillis =
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessListenerNotifier.java]: Typo still here, "ealpsed" is not a word :)

File: src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java (hunk 5 above)
Comment (afabiani) [reply src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java]: doh!

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessListenerNotifier.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (afabiani) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessListenerNotifier.java]: @aaime fixed!

File: src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java (hunk 5 above)
Comment (afabiani) [reply src/extension/wps/web-wps/src/main/java/org/geoserver/wps/web/ProcessStatusProvider.java]: @aaime fixed!

**Comment by afabiani** [comment]
@aaime can we close this one? Any other comment?

**Comment by aaime** [comment]
Only one... I'd merge it instead of closing it ;-)

**Comment by afabiani** [comment]
@aaime :P

**Pull request merged**